### PR TITLE
Request For Comments: TensorGraph

### PR DESCRIPTION
**Pull request by lilleswing**
Initial Implementation of "TensorGraph" which extends Model.

The Goal of TensorGraph is to allow for arbitrary graph execution when the graph does not have cycles.
To show how it works I train the keras-molecule autoencoder using the new model.

I'm going to continue to fill it out the class
```bash
save
reload
predict_on_batch
fit_on_batc
etc
```

The good news is that with consistent naming of layers, layers will have consistent variable naming inside their scope.  This means that we can load tf Saves even when we change our network architecture, as long as we keep some layers and also keep their names.

### CODE REVIEW

File: deepchem/models/tf_new_models/tensor_graph.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+      f = tf.Variable(
+          tf.random_normal([self.width, parent_channel_size, self.out_channels
+                           ]))
+      b = tf.Variable(tf.random_normal([self.out_channels]))
+      t = tf.nn.conv1d(parent.out_tensor, f, stride=1, padding="SAME")
+      t = tf.nn.bias_add(t, b)
+      self.out_tensor = tf.nn.relu(t)
+
+
+class Dense(Layer):
```
Comment (rbharath) [inline deepchem/models/tf_new_models/tensor_graph.py]: We should make this `Dense` layer and existing `Dense` layer in `dc.nn.copy` (or `dc.nn.layers`, forget which) agree and then get rid of one.

File: deepchem/models/tf_new_models/tensor_graph.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
+import networkx as nx
+import tensorflow as tf
+import time
+
+from deepchem.models.models import Model
+from nn.copy import Layer
+
+
+class TensorGraph(Model):
```
Comment (rbharath) [inline deepchem/models/tf_new_models/tensor_graph.py]: I'd vote for making `Sequential`, `SequentialGraph`, `SequentialSupportGraph`, `MultitaskGraphClassifier` and `MultitaskGraphRegressor` inherit from `TensorGraph`. We now have a golden opportunity for consolidating our `Tensorflow` models :-)

File: deepchem/models/tf_new_models/tensor_graph.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+          print('Ending epoch %d: Average loss %g' % (epoch, avg_loss))
+          # Always save a final checkpoint when complete.
+          saver.save(sess, self.data_dir, global_step=epoch + 1)
+      ############################################################## TIMING
+      time2 = time.time()
+      print("TIMING: model fitting took %0.3f s" % (time2 - time1))
+      ############################################################## TIMING
+
+
+class Conv1DLayer(Layer):
```
Comment (rbharath) [inline deepchem/models/tf_new_models/tensor_graph.py]: We should move these into `dc.nn.layers` or `dc.nn.copy` (`copy` indicates "copied" over from keras. We may just want to merge `dc.nn.layers` and `dc.nn.copy` together)

File: examples/encoders/encoder_example.py [hunk 4]
```diff
@@ -1,19 +1,134 @@
...
+             out_channels=gru_channels,
+             batch_size=batch_size,
+             name="GRU2")
+  gru3 = GRU(n_hidden=gru_hidden,
+             out_channels=gru_channels,
+             batch_size=batch_size,
+             name="GRU3")
+  ts_dense = TimeSeriesDense(out_channels=47)
+
+  graph_model.add_layer(input_layer, parents=list())
```
Comment (rbharath) [inline examples/encoders/encoder_example.py]: Here's a suggestion for cutting down duplication. We could have `add_layer` return the added layer.

```
input_layer = graph_model.add_layer(Input((batch_size, 120, 47), name="features"), parents=[])
conv1 = graph_model.add_layer(Conv1DLayer(width=9, out_channels=9, name="conv1"), parents=[input_layer]))
...
```

Comment (rbharath) [inline examples/encoders/encoder_example.py]: I'm not sure if this is actually neater than the way you have it though :-)

**Comment by rbharath** [comment]
@lilleswing Thanks for taking this on! Highly appreciated. I think it will add a lot of value to DeepChem by making it easier for us to build complicated models. I've made some preliminary comments on the design in the review.

File: deepchem/models/tf_new_models/tensor_graph.py (hunk 2 above)
Comment (lilleswing) [inline deepchem/models/tf_new_models/tensor_graph.py]: Yeah I like that

File: deepchem/models/tf_new_models/tensor_graph.py (hunk 3 above)
Comment (lilleswing) [inline deepchem/models/tf_new_models/tensor_graph.py]: That is a question -- I created these because they were easier for me to understand.  I think at a certain level their counterparts in dc.nn.copy do the same thing but with more bells and whistles.  The question is should we spend time to understand now?  Or when we need the bells and whistles look through keras source and bring them back.

Option 1: Understand all the functionality in counterparts of dc.nn.copy
pros:  Off the shelf regularization, momentum etc
cons: I don't understand their "call" nomenclature.  Why are layers callable?  Do all of these things pickle well?  Will there be strange interactions?

Option 2: Use the bare bones ones here and add as we need it.
pros: Understandable
cons: More time if we want things like regularization.

I personally like Option 2 -- we aren't using the features from the additional complexity in dc.nn.copy.

File: deepchem/models/tf_new_models/tensor_graph.py (hunk 1 above)
Comment (lilleswing) [inline deepchem/models/tf_new_models/tensor_graph.py]: +1 see above

File: examples/encoders/encoder_example.py (hunk 4 above)
Comment (lilleswing) [inline examples/encoders/encoder_example.py]: add_layer already adds the additional layer -- I thought it was cleaner if we didn't use it in the example :).  I actually use that syntax for the loss layer.

File: deepchem/models/tf_new_models/tensor_graph.py (hunk 3 above)
Comment (rbharath) [inline deepchem/models/tf_new_models/tensor_graph.py]: @lilleswing and I chatted about this offline. Decided to use a bare-bones version, but one which uses `__call__` in order to make layers usable by people who don't want to use full DeepChem framework (and just want to write raw Tensorflow).

**Comment by lilleswing** [comment]
Closed in favor of completed Tensorgraph model here.

**Comment by lilleswing** [comment]
https://github.com/deepchem/deepchem/pull/493
